src/selectors/print.js

```diff
@@ -13,14 +13,30 @@ const getPhotoExpanded = (photo, metadata, notes, notepad, ontology) => (
       props: ontology.props,
       template: ontology.template[photo.template]
     }),
-    notes: photo.notes.map(id => ({
-      id,
-      ...serialize(notes[id], { format: { html: true }, localize: false }),
-      ...notepad[id]
-    }))
+    notes: getPhotoNotes(photo, notepad, notes)
   }
 )
 
+const getPhotoNotes = (photo, notepad, notes) => {
+  let photoNotes = photo.notes.map(id => ({
+    id,
+    ...serialize(notes[id], { format: { html: true }, localize: false }),
+    ...notepad[id]
+  }))
+
+  let notesArr = Object.keys(notes).map(key => notes[key])
```

Comment:
This is problematic. `Object.keys` loops through all notes that exist, map then loops through all the notes one more time. You can achieve the same thing using `Object.values` or `Object.entries` with a single iteration. But should not need a note array here in the first place. Finding a note in the array below is much worse than just looking up the note via the notes object / hashtable that we already have in `notes`!

Reply:
All we want to do is loop through all the photo's notes and all its selections' notes. But this loops multiple times through all the notes of all photos and selections.

If we say that it's OK to loop through the pertinent notes twice (mapping to ids first and then to the note objects for serialization) this becomes very concise and easy to understand: `[...photo.notes, ...photo.selections.map(s => notes)].flatMap(id => ({ id, ...serialize(notes[id], ...), ....}))`. Looping through a short list of notes twice for easier legibility is totally fine.

Reply:
We'll have to pass `state.selections` to the `getPhotoExpanded`. But we really want to do that, because getting at the notes via the selections is way more efficient than the reverse of searching through all the notes for the selections.